### PR TITLE
improve migration api error handling

### PR DESCRIPTION
**Pull request by dani-mp**
Resolves: 
fist part of https://linear.app/prismic/issue/DT-2810/migration-api-client-fix-error-forwarding-coming-from-migration-api

### Description

This outlines the complete response structure for all Migration API error scenarios.

**400 Bad Request:**
- **Type:** JSON (Validation Errors Array)
```json
[
  {
    "property": "string",
    "value": "unknown", 
    "error": "string"
  }
]
```
- **Type:** Text
  - "The UID is required for this custom type"
  - "The custom type does not support UID"
  - "A document with this UID already exists"
  - "The language you provided is invalid, it must exist in your repository"
  - "There is already an existing translation in this language for the document referenced in 'alternate_language_id'"
  - "Your migration release has reached the limit of {limit} documents. Go to your repository in Prismic.io and publish your release before making another request"
  - "The `json` payload provided is malformed. {error}"
  - "Missing payload"

**401 Unauthorized:**
- **Type:** Text
- **Message:** "Unauthorized, invalid or missing authentication token"

**403 Forbidden:**
- **Type:** JSON
```json
{
  "Message": "User is not authorized to access this resource with an explicit deny"
}
```
- **Type:** Text
  - "You repository is currently in maintenance, please try again later"
  - "You repository is currently locked, please reach out to us"
  - "You repository is currently being migrated, please try again later"

**404 Not Found:**
- **Type:** Text
  - "The requested custom type does not exist"
  - "alternate_language_id: It seems that the document id you provided doesn't exist"

**429 Too Many Requests:**
- **Type:** Text (with retry-after header)
- **Message:** Rate limit exceeded

**500 Internal Server Error:**
- **Type:** Text
- **Message:** "An Unexpected error happened on our side, please contact our support"

#### Key Notes

- **Multiple Error Sources:** The API can return errors from different services (Core API, Custom Types API, validation layer) which are forwarded as-is or mapped to appropriate HTTP status codes.

- **Validation Errors:** The 400 responses can be either a structured JSON array of validation errors or simple text messages depending on the validation layer that failed.

- **Authentication/Authorization:** Different auth scenarios (invalid token, wrong repository, insufficient permissions) return different error formats.

- **Rate Limiting:** 429 responses include a `retry-after` header with the retry delay.

- **Repository State:** The API can return 403 errors when the repository is in maintenance, locked, or being migrated.

- **Document State:** Update operations can fail if the document doesn't have the required state (draft/published version).

### Checklist

- [ ] A comprehensive Linear ticket, providing sufficient context and details to facilitate the review of the PR, is linked to the PR.
- [ ] If my changes require tests, I added them.
- [ ] If my changes affect backward compatibility, it has been discussed.
- [ ] If my changes require an update to the CONTRIBUTING.md guide, I updated it.

### How to QA [^1]

[^1]:
	Please use these labels when submitting a review:
	:question: #ask:&ensp;Ask a question.
	:bulb: #idea:&ensp;Suggest an idea.
	:warning: #issue:&ensp;Strongly suggest a change.
	:tada: #nice:&ensp;Share a compliment.

### CODE REVIEW

File: src/WriteClient.ts [hunk 1]
```diff
@@ -800,12 +800,8 @@ export class WriteClient<
 
 				return { id: json.id }
 			}
-			case 403: {
-				const json = await response.json()
-				throw new ForbiddenError(json.Message, url.toString(), json)
-			}
 			default: {
-				throw new PrismicError(undefined, url.toString(), await response.text())
+				return await this.#handleMigrationAPIError(response, url)
```
[the remark this answers is not in the document]
Comment (dani-mp) [reply src/WriteClient.ts:804]: i see it more like handling everything that is not handled in the caller - everything else is an actual error or an unknown error, i.e. we don't expect a 204 here either, for instance

File: src/WriteClient.ts [hunk 2]
```diff
@@ -893,4 +881,74 @@ export class WriteClient<
...
+		response: ResponseLike,
+		url: URL,
+	): Promise<never> {
+		switch (response.status) {
+			case 400: {
+				const text = await response.text()
+				try {
+					// Try to parse as JSON validation errors array
+					const json = JSON.parse(text)
+					// `json` has the shape Array<{ property: string, value: unknown, error: string }>
```
[the remark this answers is not in the document]
Comment (dani-mp) [reply src/WriteClient.ts:906]: i've seen that we don't do this anywhere else, and when we need the data we just cast

an earlier version of this code had the casting, but as i don't need to go into the array elements, i don't think it is needed

if we try to verify the array here, what do we throw in the else? that should never happen, and in other places we're not parsing the shape anyway to be sure either

File: src/WriteClient.ts [hunk 3]
```diff
@@ -893,4 +881,74 @@ export class WriteClient<
...
+				}
+			}
+			case 404: {
+				const text = await response.text()
+				throw new NotFoundError(text, url.toString(), text)
+			}
+			case 500: {
+				const text = await response.text()
+				throw new PrismicError(text, url.toString(), text)
+			}
```
[the remark this answers is not in the document]
Comment (dani-mp) [reply src/WriteClient.ts:947]: i can totally do that, i wanted to be explicit about the status codes we do know can happen and leave the default for the ones that shouldn't

if tomorrow we want to change this, we can easily have two different errors, one for the 500 (internal error) and one for the default (unknown), for instance, and separating them here reminds the dev to take both into account differently

File: src/WriteClient.ts [hunk 4]
```diff
@@ -893,4 +881,74 @@ export class WriteClient<
...
+	 * parsing.
+	 *
+	 * @param response - The HTTP response from the Migration API.
+	 * @param url - The URL that was requested.
+	 *
+	 * @throws {@link PrismicError} For 400, 500, and other unexpected errors.
+	 * @throws {@link ForbiddenError} For 401 and 403 errors.
+	 * @throws {@link NotFoundError} For 404 errors.
+	 */
+	async #handleMigrationAPIError(
```
[the remark this answers is not in the document]
Comment (dani-mp) [reply src/WriteClient.ts:896]: we have some e2e tests using the migration api in the migration api repository

I'll use the new client there but we don't cover all the possible edge cases described in this pr, so if we want to be exhaustive for now we need to test manually

File: src/WriteClient.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (angeloashmore) [reply src/WriteClient.ts:804]: In that case, should we support only the explicit valid case?

```ts
```

File: src/WriteClient.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (angeloashmore) [reply src/WriteClient.ts:906]: As long as we are confident that the API won't send a 400 status with valid JSON for any other case, then validating the shape isn't necessary.

> if we try to verify the array here, what do we throw in the else?

It would be considered an unknown error and would fall back to `PrismicError` (unless we know the error and explicitly handle it).

File: src/WriteClient.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (angeloashmore) [reply src/WriteClient.ts]: What do you think of this?

```ts
			case 500:
			default: {
				const text = await response.text()
				throw new PrismicError(text, url.toString(), text)
			}
```

It still documents how a 500 status code is handled while eliminating the duplicated code. If we want to treat the error differently in the future, it's simple to change the code.

File: src/WriteClient.ts [hunk 5]
```diff
@@ -893,4 +881,69 @@ export class WriteClient<
...
+			case 400: {
+				try {
+					// Try to parse as JSON validation errors array
+					const json = JSON.parse(text)
+					// `json` has the shape Array<{ property: string, value: unknown, error: string }>
+					throw new PrismicError(
+						"Validation failed, check the response property of the error for details",
+						url.toString(),
+						json,
+					)
```
Comment (angeloashmore) [inline src/WriteClient.ts:911]: ⚠️ #issue: Following what was shared in a separate comment, a custom error (e.g. `InvalidDataError`) makes more sense.

In general, `PrismicError` shouldn't be used directly except as a fallback. A custom error allows the developer to determine the kind of error without parsing an error's `message`.

**Review by angeloashmore — APPROVED** [review]
Looks good! I left a tiny suggestion, but feel free to merge this + #395 and release when you're ready.
